src/metaopt/worker/experiment.py

```diff
@@ -0,0 +1,396 @@
+                            "Experiment forks will be supported soon.", len(config), name, user)
+            config = sorted(config, key=lambda x: x['metadata']['datetime'],
+                            reverse=True)[0]
+            for attrname in self.__slots__:
+                if not attrname.startswith('_'):
+                    setattr(self, attrname, config[attrname])
+            self._id = config['_id']
+
+        self._last_fetched = self.metadata['datetime']
+
+    def reserve_trial(self, score_handle=None):
+        """Find *new* trials that exist currently in database and select one of
+        them based on the highest score return from `score_handle` callable.
+
+        :param score_handle: A way to decide which trial out of the *new* ones to
+           to pick as *reserved*, defaults to a random choice.
+        :type score_handle: callable
+
+        :return: selected `Trial` object, None if could not find any.
+        """
+        if score_handle is not None and not callable(score_handle):
+            raise ValueError("Argument `score_handle` must be callable with a `Trial`.")
+
+        query = dict(
+            experiment=self._id,
+            status='new'
+            )
+        new_trials = Trial.build(self._db.read('trials', query))
+
+        if not new_trials:
+            return None
+
+        if score_handle is None:
+            selected_trial = random.sample(new_trials, 1)[0]
+        else:
+            raise NotImplementedError("scoring will be supported in the next iteration.")
+
+        selected_trial.status = 'reserved'
+        selected_trial.start_time = datetime.datetime.utcnow()
+
+        self._db.write('trials', selected_trial.to_dict(),
+                       query={'_id': selected_trial.id})
+
+        return selected_trial
+
+    def push_completed_trial(self, trial):
+        """Inform database about an evaluated `trial` with results.
+
+        :param trial: Corresponds to a successful evaluation of a particular run.
+        :type trial: `Trial`
+
+        .. note:: Change status from *reserved* to *completed*.
+        """
+        trial.end_time = datetime.datetime.utcnow()
+        trial.status = 'completed'
+        self._db.write('trials', trial.to_dict(), query={'_id': trial.id})
+
+    def register_trials(self, trials):
+        """Inform database about *new* suggested trial with specific parameter
+        values. Each of them correspond to a different possible run.
+
+        :type trials: list of `Trial`
+        """
+        stamp = datetime.datetime.utcnow()
+        for trial in trials:
+            trial.experiment = self._id
+            trial.status = 'new'
+            trial.submit_time = stamp
+        trials_dicts = list(map(lambda x: x.to_dict(), trials))
+        self._db.write('trials', trials_dicts)
+
+    def fetch_completed_trials(self):
+        """Fetch recent completed trials that this `Experiment` instance has not
```

Comment:
I believe that we should not buffer completed trials. Memory is going to blow up. Algorithms can use incrementally observed completed trials to update their internal state. But should we want we must be able to fetch all of them.

What would be the cases in which we want to have all completed trials in a list?

Reply:
in experiment.stats for instance.

Reply:
I do not think it will be a good idea memory-wise. Even later, I think that we should fetch all completed trials in batches and calculate the stats on them, should the list become too large. What do you think?

Reply:
Yes, I agree.